plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java

```diff
@@ -61,11 +61,9 @@ public Object compute(IEclipseContext context, String contextKey) {
 
     Object path = context.get(PreferenceConstants.CLOUDSDK_PATH);
     File location = toFile(path);
-    CloudSdk.Builder builder = new CloudSdkProvider(null).createBuilder(location);
-    if (builder == null) {
-      return NOT_A_VALUE;
-    }
-    CloudSdk instance = builder.build();
+    // TODO(joaomartins): When can the SDK location in the context be different from the one in the
+    // preference store?
+    CloudSdk instance = new CloudSdkProvider().getCloudSdk();
```

Comment:
Namely: singletons are evil.  Look at your patch where you have several uses of `new CloudSdkProvider().getCloudSdk()` — you're assuming that subsequent invocations produce an equivalent instance.  We should avoid baking in assumptions that CloudSdk instances are always from the same path.

In this case, the Context Functions form part of the Eclipse 4 Dependency Injection system.  The Eclipse Contexts are a bit like [Guice modules on steroids](https://www.eclipsecon.org/na2014/sites/default/files/slides/2014-EclipseCon-DI.pdf): they're very dynamic, and accesses to variables can be computed based on other elements found within the context (like what's happening here).  The contexts remember what variables are looked up from a context and reruns the computations whenever one of those variables change.

Put differently, we need some variable access to trigger recompilation on a change, and the sdkPath seems a logical element to trigger it.